graph/filters.py

```diff
@@ -0,0 +1,106 @@
+"""Filters for usage in pathfinding."""
+
+from dataclasses import dataclass
+from typing import Any, Callable, Iterable, Union
+
+EdgeData = tuple[str, str, dict]
+EdgeList = Iterable[EdgeData]
+
+EdgeDataExtractor = Callable[[EdgeData, Any], Any]
+
+@dataclass
+class UseValIfNone:
+    processor: Callable[[EdgeData], Any]
+
+    def __call__(self, edge: EdgeData, val: Any):
+        result = self.processor(edge)
+        if result is None:
+            return val
+        return result
+    
+@dataclass
+class UseDefaultIfNone:
+    processor: Callable[[EdgeData], Any]
+    default: Any
+
+    def __call__(self, edge: EdgeData, _):
+        result = self.processor(edge)
+        if result is None:
+            return self.default
+        return result
+
+@dataclass
+class ProcessEdgeAttribute:
+    attribute: str
+    processor: Callable[[Any], Any] = None
+
+    def __call__(self, edge: EdgeData):
+        result = edge[2].get(self.attribute)
+        if self.processor:
+            return self.processor(result)
+        return result
+
+class EdgeFilter:
+
+    def __init__(
+        self,
+        comparator: Callable[[Any, Any], bool],
+        get_func: Union[EdgeDataExtractor, str],
+        preprocessor: Callable[[Any], Any] = None
+    ):
+        self.comparator = comparator
+        
+        if isinstance(get_func, str):
+            get_func = UseValIfNone(ProcessEdgeAttribute(get_func))
+
+        self.get_func = get_func
+
+        self.preprocessor = preprocessor
+
+    def __call__(
+        self,
+        value: Any,
+        items: EdgeList
+    ) -> EdgeList:
+        """Apply the filter given the items and value."""
+
+        # Preprocess the value
+        if self.preprocessor:
+            value = self.preprocessor(value)
+        
+        # Run filter
+        return filter(
+            lambda edge:
+                self.comparator(
+                    self.get_func(edge, value),
+                    value
+            ),
+            items
+        )
+
+@dataclass
+class TypeCheckPreprocessor:
+    types: Union[type, tuple[type]]
+
+    def __call__(
+        self,
+        value: Any
+    ):
+        if not isinstance(value, self.types):
+            raise TypeError(f"Expected type: {self.types}")
+        return value
+
+@dataclass
+class TypeDifferentiatedProcessor:
+    preprocessors: dict[Union[type, tuple[type]], Callable[[Any], Any]] = None
+
+    def __call__(
+        self,
+        value: Any
+    ):
+        for expected_types, processor in self.preprocessors.items():
+            if isinstance(value, expected_types):
+                if processor:
+                    return processor(value)
+                return value
+        raise TypeError(f"Expected types: {self.preprocessors.keys()}")
```

Comment:
```
❯ curl -s -X POST -H 'Content-type: application/json' http://127.0.0.1:8181/api/kytos/pathfinder/v3/ -d '{"source": "00:00:00:00:00:00:00:01:1", "destination":  "00:00:00:00:00:00:00:03:
1", "spf_max_path_cost": 10, "mandatory_metrics": {"ownership": "yellow"}}' | jq
{
  "description": "Expected types: dict_keys([<class 'str'>, <class 'dict'>, <class 'NoneType'>])",
  "code": 400
}
```

If an edge metadata were to be improperly configured with a type, the current error message is somewhat helpful (especially for folks who understand how it's implemented), but it's not that easy to understand what the error was. Can you try to make this error message with a bit more context just so network operators might be able to easily figure out if it ever happens? Maybe also including the `value` and its type in the error?

Reply:
Added in some more information here. It should now include the specific edge, attribute, and the repr of the value of the attribute.